init/cli/status.go

```diff
@@ -35,27 +38,90 @@ If exit code is nonzero, writes an error message to stderr and var/log/startup.l
+				"having been started", commandNames(serviceStatus.notRunningCmds), serviceStatus.writtenPids)
+		},
+	}
+	NotRunning = ServiceState{
+		Description: "Service not running",
+		Applicable: func(serviceStatus *serviceStatus, err error) bool {
+			return err == nil && len(serviceStatus.notRunningCmds) > 0 && len(serviceStatus.writtenPids) == 0
+		},
+		ExitStatus: func(serviceStatus *serviceStatus, err error) (int, error) {
+			return 3, errors.Errorf("commands '%v' are not running", commandNames(serviceStatus.notRunningCmds))
+		},
+	}
+	ErrorState = ServiceState{
+		Description: "Failed to determine service status",
+		Applicable: func(serviceStatus *serviceStatus, err error) bool {
+			return err != nil
+		},
+		ExitStatus: func(serviceStatus *serviceStatus, err error) (int, error) {
+			if err != nil {
+				return 4, errors.Wrap(err, "failed to determine service status")
+			}
+			return 4, errors.Errorf("failed to determine service status")
+		},
+	}
+)
+
 func status(ctx cli.Context, loggers launchlib.ServiceLoggers) error {
 	// Executed with logging for errors, however we discard the verbose logging of getServiceStatus
 	serviceStatus, err := getServiceStatus(ctx, &DevNullLoggers{})
-	if err != nil {
-		return logErrorAndReturnWithExitCode(ctx, errors.Wrap(err, "failed to determine service status"), 4)
-	}
-	if len(serviceStatus.notRunningCmds) > 0 {
-		notRunningCmdNames := make([]string, 0, len(serviceStatus.notRunningCmds))
-		for name := range serviceStatus.notRunningCmds {
-			notRunningCmdNames = append(notRunningCmdNames, name)
+	var matched *ServiceState
+	for _, state := range []ServiceState{ErrorState, NotRunning, Dead, Running} {
+		if state.Applicable(serviceStatus, err) {
+			matched = &state
+			break
 		}
-		if len(serviceStatus.writtenPids) > 0 {
-			return logErrorAndReturnWithExitCode(
-				ctx,
-				errors.Errorf("commands '%v' are not running but there is a record of commands '%v'"+
-					"having been started", notRunningCmdNames, serviceStatus.writtenPids),
-				1,
-			)
+	}
+
+	// If no state has matched, default to error state
+	if matched == nil {
+		matched = &ErrorState
+	}
+
+	code, err := matched.ExitStatus(serviceStatus, err)
+	if code != 0 {
+		fmt.Fprintln(os.Stderr, matched.Description)
+		if err != nil {
+			return logErrorAndReturnWithExitCode(ctx, err, code)
 		}
-		return logErrorAndReturnWithExitCode(ctx, errors.Errorf("commands '%v' are not running",
-			notRunningCmdNames), 3)
+		// Non-zero exit codes can only be reported through cli.WithExitCode which must include an errors, though that
+		// error can be empty.  Returning no error gives an exit code of 0
+		return cli.WithExitCode(code, errors.New(""))
 	}
+
+	fmt.Println(matched.Description)
 	return nil
 }
+
+type ServiceState struct {
+	Description string
+	Applicable  func(serviceStatus *serviceStatus, err error) bool
+	ExitStatus  func(serviceStatus *serviceStatus, err error) (int, error)
+}
```

Comment:
Is this coupling of data + methods into a struct idiomatic go?

I'd prefer to have this be an interface instead.

Reply:
I don't want to speak for the go community but I dislike the interface approach since that forces the creation of separate four separate types for each `ServiceState`, all of them being stateless singletons.

Reply:
As a slightly less hand-wavey comment, I think that while there is data and a method, the methods are conceptually related to the data, but none of their logic is.  In some ways the point is that these functions are in fact functions and strictly not methods, but it's useful to have a construct which ties them too each other, similar to a named tuple, since a `State` involves a particular description, definition of whether its applicable, and generation of exit code.

Might be easier to talk over in person if that isn't clear.